### PR TITLE
new(falco): add append_output configuration option with fields and format

### LINKED ISSUE #3235: [proposal/discussion] specify Falco additional outputs in the configuration file
**Motivation**

There are several issues open for providing more flexibility for the Falco output format:
see:
* https://github.com/falcosecurity/falco/issues/2127
* https://github.com/falcosecurity/falco/issues/2289

And I might be missing some.

Basically, the ask is:
* Add more things to the rule output in the message (e.g. add "cluster name: %some_value"). This of course is specific by Falco setup
* Add more things to the rule output in the json fields but not in the message, so it won't pollute the output string but would be parseable by any log aggregator
* Ability to select which source (syscall/plugin) and/or which _rule tag_ is affected by a specific change

**Feature**

I'm thinking about how to solve this and I came up with this option in Falco.yaml

```yaml
add_output:
  - source: syscall
    tag: persistence
    format: "gparent=%proc.aname[2] ggparent=%proc.aname[3] gggparent=%proc.aname[4]"
  - source: k8s_audit
    fields: ["ka.verb", "ka.uri"]
    format: "with response code %ka.response.code
```

What I do like about this is that it allows to do it all, select only specific tags, specify both format and fields per source. What I don't like too much is that it makes it hard to specify it on the command line with `-o`. Comments welcome.

### PR DESCRIPTION
**Pull request by LucaGuerra**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area engine

**What this PR does / why we need it**:

This addresses the need for a configuration option that allows to specify:
1. Additional output at the end of the regular message
2. Additional output fields, only visible in the json message, which can contain any custom message including formatted fields. Supports outputting strings with format and environment variables.

This is how it works

```yaml
append_output:
  - source: syscall
    tag: persistence
    rule: some rule name
    format: "gparent=%proc.aname[2] ggparent=%proc.aname[3] gggparent=%proc.aname[4]"

  - source: k8s_audit
    fields: 
      - ka.verb
      - home_directory: "${HOME}"
      - my_field: "this is event number %evt.num"
    format: "with response code %ka.response.code"
```

**Which issue(s) this PR fixes**:

Fixes #3235

**Special notes for your reviewer**:

As a bonus, this gets rid of JsonCpp in Falco.

**Does this PR introduce a user-facing change?**:

```release-note
new(falco): introduce append_output configuration
```

### CODE REVIEW

**Comment by LucaGuerra** [comment]
The structure and functionality is there and can be reviewed, I need to figure out how to properly add this to the json schema and add an explanation to `falco.yaml`

**Comment by LucaGuerra** [comment]
I have updated the PR:
* The json schema should be updated to reflect these changes
* There is an explanation in `falco.yaml` about how to use this feature
* Fields specified as `- ka.verb` are now supported. They will show up with their original type instead of being cast to string

**Comment by LucaGuerra** [comment]
This fixes the `falcosecurity/testing` failure: https://github.com/falcosecurity/testing/pull/61

File: falco.yaml [hunk 1]
```diff
@@ -576,6 +576,42 @@ rule_matching: first
...
+# And several options to add output:
+#   `format`: add output to the Falco message
+#   `fields`: add new fields to the JSON output and structured output, which will not
+#             affect the regular Falco message in any way. These can be specified as a
+#             custom name with a custom format or as any supported field
+#             (see: https://falco.org/docs/reference/rules/supported-fields/)
+#
+# Example:
+#
+#   - source: syscall
```
Comment (FedeDP) [inline falco.yaml:604]:
```suggestion
# append_output:
#   - source: syscall
```

File: userspace/engine/falco_engine.h [hunk 2]
```diff
@@ -196,6 +221,7 @@ class falco_engine
 		std::string format;
 		std::set<std::string> exception_fields;
 		std::set<std::string> tags;
+		std::unordered_map<std::string, std::pair<std::string, bool>> extra_output_fields;
```
Comment (FedeDP) [inline userspace/engine/falco_engine.h:224]: I saw this is written many times, can we `typedef` it to a meaningful name?
Like:
```
typedef extra_output_field_t std::unordered_map<std::string, std::pair<std::string, bool>>
```

File: userspace/engine/formats.cpp [hunk 3]
```diff
@@ -15,7 +15,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-#include <json/json.h>
+#include <nlohmann/json.hpp>
```
Comment (FedeDP) [inline userspace/engine/formats.cpp:18]: Thanks for this :laughing:

File: userspace/engine/formats.cpp [hunk 4]
```diff
@@ -94,43 +84,54 @@ std::string falco_formats::format_event(sinsp_evt *evt, const std::string &rule,
...
-
-		// Json::FastWriter may add a trailing newline. If it
-		// does, remove it.
-		if(full_line[full_line.length() - 1] == '\n')
-		{
-			full_line.resize(full_line.length() - 1);
+			if(ef.second.second) // raw field
+			{
+				std::string json_field_map;
+				formatter = m_falco_engine->create_formatter(source, fformat);
```
Comment (FedeDP) [inline userspace/engine/formats.cpp:108]: I am wondering whether `create_formatter` is a slow process; perhaps we might cache formatters per `<source,fformat>` key pair?

File: userspace/engine/falco_engine.h (hunk 2 above)
Comment (LucaGuerra) [reply userspace/engine/falco_engine.h]: Thanks! Done.

**Comment by LucaGuerra** [comment]
Addressed @FedeDP 's comment, updated the json schema and also added a test for CLI options since the `-o` PR was merged in the meantime

**Review by FedeDP — APPROVED** [review]
/approve

**Comment by leogr** [comment]
> This PR may bring feature or behavior changes in the Falco engine and may require the engine version to be bumped.
> 
> Please double check **userspace/engine/falco_engine_version.h** file. See [versioning for FALCO_ENGINE_VERSION](https://github.com/falcosecurity/falco/blob/master/RELEASE.md#falco-repo-this-repo).
> 
> /hold

false positive
/unhold

**Review by leogr — COMMENTED** [review]
nits

**Review by leogr — APPROVED** [review]

**Pull request merged**